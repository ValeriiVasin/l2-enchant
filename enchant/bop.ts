import { Item } from '../items';
import { Enchant, EnchantMap } from '../types';

const chances = [100, 100, 100, ...Array(20).fill(66.6 + 5)];
const items = [
  Item.BOP,
  Item.BOP_1,
  Item.BOP_2,
  Item.BOP_3,
  Item.BOP_4,
  Item.BOP_5,
  Item.BOP_6,
  Item.BOP_7,
  Item.BOP_8,
  Item.BOP_9,
  Item.BOP_10,
  Item.BOP_11,
  Item.BOP_12,
  Item.BOP_13,
  Item.BOP_14,
  Item.BOP_15,
  Item.BOP_16,
];

/**
 * @TODO improve Bx calculation
 */
export const bopEnchantMap: EnchantMap = new Map(
  items.slice(0, -1).map((item, index): [Item, Enchant] => {
    return [
      item,
      {
        item,
        required: Item.EWB,
        success: items[index + 1],
        fail: Item.Nothing,
        successRate: chances[index],
      },
    ];
  })
);
