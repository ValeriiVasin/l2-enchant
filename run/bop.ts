import { enchant } from '../helpers/enchant';
import { Item } from '../types';

const enchants: Array<[number, Item]> = [
  [3, Item.BOP_3],
  [4, Item.BOP_4],
  [5, Item.BOP_5],
  [6, Item.BOP_6],
  [7, Item.BOP_7],
  [8, Item.BOP_8],
  [9, Item.BOP_9],
  [10, Item.BOP_10],
  [11, Item.BOP_11],
  [12, Item.BOP_12],
  [13, Item.BOP_13],
  [14, Item.BOP_14],
  [15, Item.BOP_15],
  [16, Item.BOP_16],
];

for (let [lvl, item] of enchants) {
  console.log(`Лук Угрозы +0 => +${lvl}`, enchant(Item.BOP, item, 1000));
}
