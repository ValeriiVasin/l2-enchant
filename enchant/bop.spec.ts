import { bopEnchantMap } from './bop';

it('enchant map: bop', () => {
  expect(bopEnchantMap).toMatchSnapshot();
});
