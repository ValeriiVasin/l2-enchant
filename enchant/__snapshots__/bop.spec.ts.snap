// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enchant map: bop 1`] = `
Map {
  "Лук Угрозы" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы",
    "results": Array [
      Object {
        "chance": 100,
        "item": "Лук Угрозы +1",
      },
      Object {
        "chance": 0,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +1" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +1",
    "results": Array [
      Object {
        "chance": 100,
        "item": "Лук Угрозы +2",
      },
      Object {
        "chance": 0,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +2" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +2",
    "results": Array [
      Object {
        "chance": 100,
        "item": "Лук Угрозы +3",
      },
      Object {
        "chance": 0,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +3" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +3",
    "results": Array [
      Object {
        "chance": 71.6,
        "item": "Лук Угрозы +4",
      },
      Object {
        "chance": 28.400000000000006,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +4" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +4",
    "results": Array [
      Object {
        "chance": 71.6,
        "item": "Лук Угрозы +5",
      },
      Object {
        "chance": 28.400000000000006,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +5" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +5",
    "results": Array [
      Object {
        "chance": 71.6,
        "item": "Лук Угрозы +6",
      },
      Object {
        "chance": 28.400000000000006,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +6" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +6",
    "results": Array [
      Object {
        "chance": 71.6,
        "item": "Лук Угрозы +7",
      },
      Object {
        "chance": 28.400000000000006,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +7" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +7",
    "results": Array [
      Object {
        "chance": 71.6,
        "item": "Лук Угрозы +8",
      },
      Object {
        "chance": 28.400000000000006,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +8" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +8",
    "results": Array [
      Object {
        "chance": 71.6,
        "item": "Лук Угрозы +9",
      },
      Object {
        "chance": 28.400000000000006,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +9" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +9",
    "results": Array [
      Object {
        "chance": 71.6,
        "item": "Лук Угрозы +10",
      },
      Object {
        "chance": 28.400000000000006,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +10" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +10",
    "results": Array [
      Object {
        "chance": 71.6,
        "item": "Лук Угрозы +11",
      },
      Object {
        "chance": 28.400000000000006,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +11" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +11",
    "results": Array [
      Object {
        "chance": 71.6,
        "item": "Лук Угрозы +12",
      },
      Object {
        "chance": 28.400000000000006,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +12" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +12",
    "results": Array [
      Object {
        "chance": 71.6,
        "item": "Лук Угрозы +13",
      },
      Object {
        "chance": 28.400000000000006,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +13" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +13",
    "results": Array [
      Object {
        "chance": 71.6,
        "item": "Лук Угрозы +14",
      },
      Object {
        "chance": 28.400000000000006,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +14" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +14",
    "results": Array [
      Object {
        "chance": 71.6,
        "item": "Лук Угрозы +15",
      },
      Object {
        "chance": 28.400000000000006,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +15" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +15",
    "results": Array [
      Object {
        "chance": 71.6,
        "item": "Лук Угрозы +16",
      },
      Object {
        "chance": 28.400000000000006,
        "item": "Nothing",
      },
    ],
  },
}
`;
